core/go/internal/sequencer/coordinator/coordinator.go

```diff
@@ -0,0 +1,600 @@
+	syncPoints         syncpoints.SyncPoints
+	readyForDispatch   func(context.Context, *transaction.Transaction)
+	coordinatorStarted func(contractAddress *pldtypes.EthAddress, coordinatorNode string)
+	coordinatorIdle    func(contractAddress *pldtypes.EthAddress)
+	heartbeatCtx       context.Context
+	heartbeatCancel    context.CancelFunc
+	metrics            metrics.DistributedSequencerMetrics
+
+	/*Algorithms*/
+	transactionSelector TransactionSelector
+
+	/* Event loop */
+	coordinatorEvents chan common.Event
+	externalQueries   chan QueryRequest
+	stopEventLoop     chan struct{}
+
+	/* Dispatch loop */
+	dispatchQueue    chan *transaction.Transaction
+	stopDispatchLoop chan struct{}
+}
+
+func NewCoordinator(
+	ctx context.Context,
+	contractAddress *pldtypes.EthAddress,
+	domainAPI components.DomainSmartContract,
+	transportWriter transport.TransportWriter,
+	clock common.Clock,
+	engineIntegration common.EngineIntegration,
+	syncPoints syncpoints.SyncPoints,
+	requestTimeout,
+	assembleTimeout common.Duration,
+	blockRangeSize uint64,
+	blockHeightTolerance uint64,
+	closingGracePeriod int,
+	maxInflightTransactions int,
+	maxDispatchAhead int,
+	nodeName string,
+	metrics metrics.DistributedSequencerMetrics,
+	readyForDispatch func(context.Context, *transaction.Transaction),
+	coordinatorStarted func(contractAddress *pldtypes.EthAddress, coordinatorNode string),
+	coordinatorIdle func(contractAddress *pldtypes.EthAddress),
+) (*coordinator, error) {
+	c := &coordinator{
+		heartbeatIntervalsSinceStateChange: 0,
+		transactionsByID:                   make(map[uuid.UUID]*transaction.Transaction),
+		domainAPI:                          domainAPI,
+		transportWriter:                    transportWriter,
+		contractAddress:                    contractAddress,
+		blockHeightTolerance:               blockHeightTolerance,
+		closingGracePeriod:                 closingGracePeriod,
+		maxInflightTransactions:            maxInflightTransactions,
+		maxDispatchAhead:                   maxDispatchAhead,
+		grapher:                            transaction.NewGrapher(ctx),
+		clock:                              clock,
+		requestTimeout:                     requestTimeout,
+		assembleTimeout:                    assembleTimeout,
+		engineIntegration:                  engineIntegration,
+		syncPoints:                         syncPoints,
+		readyForDispatch:                   readyForDispatch,
+		coordinatorStarted:                 coordinatorStarted,
+		coordinatorIdle:                    coordinatorIdle,
+		nodeName:                           nodeName,
+		metrics:                            metrics,
+		coordinatorEvents:                  make(chan common.Event, 50), // TODO >1 only required for sqlite coarse-grained locks. Should this be DB-dependent?
+		externalQueries:                    make(chan QueryRequest, 10),
+		stopEventLoop:                      make(chan struct{}),
+		coordinatorSelectionBlockRange:     blockRangeSize,
```

Comment:
Feels like there's a bunch of inputs on this function that are really a `config` and it would be easier to just copy one whole struct in so tuning could expand more easily in the future.